File: 1443-minimum-time-to-collect-all-apples-in-a-tree/1443-minimum-time-to-collect-all-apples-in-a-tree.py
class Solution:
    def minTime(self, n: int, edges: List[List[int]], hasApple: List[bool]) -> int:
        # edgesUsed={}
        # graph=defaultdict(list)
        # for u,v in edges:
        #     graph[u].append(v)
        #     graph[v].append(u)
        # def dfs(i):
        #     visited[i]=True
        #     for out in graph[src]:
        #         if visited[out]==False:
        #             dfs(out)
        # dfs(0)
#         g = collections.defaultdict(list)
#         seen = set()
        
#         for u,v in edges:
#             g[u].append(v)
#         stack=[]
#         stack.append(0)
#         visited=[False]*(n)
#         ans=0
#         while stack:
#             aux=stack.pop()
#             visited[aux]=True
#             for out in g[aux]:
#                 if visited[out]==False:
#                     stack.append(out)
#                 if hasApple[out]==True:
#                     ans+=2
#         return ans        
#         def dfs(node):
#             seen.add(node)
#             res = 0
            
#             for nei in g[node]:
#                 if nei not in seen:
#                     res+= dfs(nei)
            
#             if ( res>0 or hasApple[node]) and node!=0:
#                 res+=2
            
#             return res
        
       
            
        # return dfs(0)
        hasApple[0], degree = True, [0] * n
        graph = collections.defaultdict(list)
        for u, v in edges:
            graph[u].append(v)
            graph[v].append(u)
            degree[u] += 1
            degree[v] += 1
        queue = collections.deque(v for v in range(n) if degree[v] == 1)
        while queue:
            u = queue.popleft()
            if hasApple[u]: continue
            for v in graph[u]:
                if degree[v] > 0:
                    degree[v] -= 1
                    degree[u] -= 1
                    if degree[v] == 1:
                        queue.append(v)
        print(degree)                
        return sum(degree)
        
        
        

File: 1443-minimum-time-to-collect-all-apples-in-a-tree/1443-minimum-time-to-collect-all-apples-in-a-tree.cpp
class Solution {
public:
    int dfs(int src,int parent,vector<int> adj[],vector<bool> &hasApple ){
        int total_time=0;
        int child_time=0;
        for(auto child: adj[src]){
            if(child==parent) continue;
            child_time=dfs(child,src,adj,hasApple);
            if(child_time || hasApple[child]) total_time+=(child_time+2);
        }
        return total_time;
    }
    int minTime(int n, vector<vector<int>>& edges, vector<bool>& hasApple) {
        vector<int> adj[n];
        for(auto &edge: edges){
            adj[edge[0]].push_back(edge[1]);
            adj[edge[1]].push_back(edge[0]);
        }
        return dfs(0,-1,adj,hasApple);
        
    }
};